src/shapes.nim

```diff
@@ -0,0 +1,277 @@
+        return newPoint2D(u, arccos(pt.z) / PI)
+
+    of skPlane: 
+        return newPoint2D(pt.x - floor(pt.x), pt.y - floor(pt.y))
+
+
+proc normal*(shape: Shape; pt: Point3D, dir: Vec3f): Normal = 
+    case shape.kind
+    of skAABox:
+        let aabb = shape.aabb.get
+        if   pt.x == aabb.min.x or pt.x == aabb.max.x: result = newNormal(1, 0, 0)
+        elif pt.y == aabb.min.y or pt.y == aabb.max.y: result = newNormal(0, 1, 0)
+        elif pt.z == aabb.min.z or pt.z == aabb.max.z: result = newNormal(0, 0, 1)
+        else: quit "Something went wrong in calculating the normal for an AABox."
+        return sgn(-dot(result.Vec3f, dir)).float32 * result
+
+    of skTriangle:
+        let 
+            (A, B, C) = shape.vertices
+            cross = cross((B - A).Vec3f, (C - A).Vec3f)
+        return sgn(-dot(cross, dir)).float32 * cross.toNormal
+        
+    of skMesh: discard
+
+    of skSphere: 
+        return sgn(-dot(pt.Vec3f, dir)).float32 * newNormal(pt.x, pt.y, pt.z)
+
+    of skPlane: 
+        return newNormal(0, 0, sgn(-dir[2]).float32)
+
+
+proc fastIntersection*(shape: Shape, ray: Ray): bool =
+    case shape.kind
+    of skAABox:
+        let 
+            (min, max) = (shape.min - ray.origin, shape.max - ray.origin)
+            (tx_min, tx_max) = (min.x / ray.dir[0], max.x / ray.dir[0])
+            (ty_min, ty_max) = (min.y / ray.dir[1], max.y / ray.dir[1])
+    
+        if tx_min > ty_max or ty_min > tx_max: return false
+
+        let
+            t_hit_min = max(ty_min, tx_min)
+            t_hit_max = min(ty_max, tx_max)
+            (tz_min, tz_max) = (min.z / ray.dir[2], max.z / ray.dir[2])
+            
+        return if t_hit_min > tz_max or tz_min > t_hit_max: false else: true
+
+    of skMesh: discard
+    of skTriangle: discard
+
+    of skSphere: 
+        let 
+            inv_ray = apply(shape.transf.inverse, ray)
+            (a, b, c) = (norm2(inv_ray.dir), dot(inv_ray.origin.Vec3f, inv_ray.dir), norm2(inv_ray.origin.Vec3f) - 1.0)
+            delta_4 = b * b - a * c
+
+        if delta_4 < 0: return false
+
+        let (t_l, t_r) = ((-b - sqrt(delta_4)) / a, (-b + sqrt(delta_4)) / a)
+        return (inv_ray.tmin < t_l and t_l < inv_ray.tmax) or (inv_ray.tmin < t_r and t_r < inv_ray.tmax) 
+
+    of skPlane:
+        let inv_ray = apply(shape.transf.inverse, ray)
+        if abs(inv_ray.dir[2]) < epsilon(float32): return false
+
+        let t = -inv_ray.origin.z / inv_ray.dir[2]
+        return if t < inv_ray.tmin or t > inv_ray.tmax: false else: true
+
+
+
+type
+    HitRecord* = object
+        ray*: Ray
+        t_hit*: float32
+        world_pt*: Point3D
+        surface_pt*: Point2D
+        normal*: Normal
+
+proc rayIntersection*(shape: Shape, ray: Ray): Option[HitRecord] =
+    var 
+        t_hit: float32
+        hit_pt: Point3D
+        surf_pt: Point2D
+        normal: Normal
+
+    let inv_ray = apply(shape.transf.inverse, ray)
+
+    case shape.kind
+    of skTriangle:
```

Comment:
I feel like that it's not completely user friendly this part of source code, we could briefly explain what we are doing in order to make source code analysis less tiring